pkg/ring/token_checker.go

```diff
@@ -0,0 +1,253 @@
+package ring
+
+import (
+	"context"
+	"flag"
+	"sync"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+)
+
+// TokenCheckerConfig is the config to configure a TokenChecker.
+type TokenCheckerConfig struct {
+	// Used by ingester when set to true.
+	CheckOnCreate   bool `yaml:"check_on_create,omitempty"`
+	CheckOnAppend   bool `yaml:"check_on_append,omitempty"`
+	CheckOnTransfer bool `yaml:"check_on_transfer,omitempty"`
+
+	// Used by TokenChecker on the interval
+	CheckOnInterval time.Duration `yaml:"check_on_interval"`
+}
+
+// RegisterFlags adds flags required to configure a TokenChecker to
+// the provided FlagSet.
+func (cfg *TokenCheckerConfig) RegisterFlags(f *flag.FlagSet) {
+	cfg.RegisterFlagsWithPrefix("", f)
+}
+
+// RegisterFlagsWithPrefix adds the flags required to config a TokenChecker to
+// the given FlagSet, prefixing each flag with the value provided by prefix.
+func (cfg *TokenCheckerConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.BoolVar(&cfg.CheckOnCreate, prefix+"token-checker.check-on-create", false, "Check that newly created streams fall within expected token ranges")
+	f.BoolVar(&cfg.CheckOnCreate, prefix+"token-checker.check-on-append", false, "Check that existing streams appended to fall within expected token ranges")
+	f.BoolVar(&cfg.CheckOnCreate, prefix+"token-checker.check-on-transfer", false, "Check that streams transferred in using the transfer mechanism fall within expected token ranges")
```

Comment:
Oops, just noticed I'm binding the same variable for all flags. Fix incoming soon.